xcarchive/entitlements.go

```diff
@@ -0,0 +1,77 @@
+package xcarchive
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/bitrise-io/go-utils/command"
+	"github.com/bitrise-io/go-utils/env"
+	"github.com/bitrise-io/go-utils/pathutil"
+	"github.com/bitrise-io/go-xcode/plistutil"
+)
+
+func executableNameFromInfoPlist(infoPlist plistutil.PlistData) string {
+	if name, ok := infoPlist["CFBundleExecutable"].(string); ok {
+		return name
+	}
+	return ""
+}
+
+func getEntitlements(basePath, entitlementsRelativePath, executableRelativePath string) (plistutil.PlistData, error) {
+	entitlements, err := entitlementsFromXcentFile(basePath, entitlementsRelativePath)
+
+	if err != nil {
+		return plistutil.PlistData{}, err
+	}
+
+	if entitlements != nil {
+		return *entitlements, nil
+	}
+
+	entitlements, err = entitlementsFromExecutable(basePath, executableRelativePath)
+
+	if err != nil {
+		return plistutil.PlistData{}, err
+	}
+
+	if entitlements != nil {
+		return *entitlements, nil
+	}
+
+	return plistutil.PlistData{}, nil
+}
+
+func entitlementsFromXcentFile(basePath, entitlementsRelativePath string) (*plistutil.PlistData, error) {
+	entitlementsPath := filepath.Join(basePath, entitlementsRelativePath)
+	exist, err := pathutil.IsPathExists(entitlementsPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if entitlements exists at: %s, error: %s", entitlementsPath, err)
+	}
+
+	if exist == false {
+		return nil, nil
+	}
+
+	plist, err := plistutil.NewPlistDataFromFile(entitlementsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &plist, nil
+}
+
+func entitlementsFromExecutable(basePath, executableRelativePath string) (*plistutil.PlistData, error) {
+	factory := command.NewFactory(env.NewRepository())
+	cmd := factory.Create("codesign", []string{"--display", "--entitlements", ":-", filepath.Join(basePath, executableRelativePath)}, nil)
+	entitlementsString, err := cmd.RunAndReturnTrimmedOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	plist, err := plistutil.NewPlistDataFromContent(entitlementsString)
```

Comment:
On my machine the first line of the output is: `Executable=<bin_path>`, is not it causing issues at plist parsing?

Reply:
I did not even notice it until you called it out. The tests work and they compare a real entitlement map read from the executable. 

I checked the command output and doing it this way
```
factory := command.NewFactory(env.NewRepository())
cmd := factory.Create("codesign", []string{"--display", "--entitlements", ":-", filepath.Join(basePath, executableRelativePath)}, nil)
entitlementsString, err := cmd.RunAndReturnTrimmedOutput()
```
only returns the actual xml entitlements string. The `Executable=<bin_path>` is missing. I do not see yet why this returns different output compared to executing it manually in the terminal. Do you have any idea?

Reply:
Not sure if it is the case, but the codesign binary could check if it is running in an interactive terminal or not. I have seen brew CLI do that.

Reply:
Okay, too many open questions. I decided to open Hopper and have a look at the codesign internals. 

It seems like codesign can output various information like
- "Executable=%s"
- "Identifier=%s"
- "CodeDirectory v=%x size=%d flags=%s hashes=%d+%d location=%s"
- and so on

Every log seems like to have a log level. For the executable line it is 0 and something greater for all of the rest. That could be the reason why only the executable path is printed out.

The more interesting thing is that it uses the stderr to print out these informations. I think it is probably a workaround to separate the entitlements output from the verbose log messages because these logs probably should not go to the err output in normal circumstances. 

And this explains why `RunAndReturnTrimmedOutput` does not return it. I tried it with `RunAndReturnTrimmedCombinedOutput` and then the executable path was returned.

So I think we are good with this.